### PR TITLE
[Menu] [Performance] Improve Menu item can execute logic. Make it work like in WPF

### LINKED ISSUE #6434: [MenuItem] [Performance] Bad CanExecute logic
**Describe the bug**
1) CanExecute evaluation are called for Command and CommandParameter on every prop change. Adds load on startup(I've found a significant marker in profiler)
2) CanExecute evaluation are called for all menus in app on any CanExecute of any Command, even if they are not visible
3) CanExecute evaluation are not called on menu opening
This behaviors above are not the same as in WPF. This caused a lot of port bugs in our app.

How WPF are doing?
1) Do not evaluate canExecute if menu is not opened
2) Evaluated canExecutes on menu opening

You can check MenuItem code in source.dot.net and menuItem code in avalonia to find the differenced in logic.
most important thing in a code below. Even a comment says - `perf optimization`
```
private void UpdateCanExecute()
        {
            MenuItem.SetBoolField(this, BoolField.CanExecuteInvalid, false);
            if (Command != null)
            {
                // Perf optimization - only raise CanExecute event if the menu is open
                MenuItem parent = ItemsControl.ItemsControlFromItemContainer(this) as MenuItem;
                if (parent == null || parent.IsSubmenuOpen)
                {
                    CanExecute = MS.Internal.Commands.CommandHelpers.CanExecuteCommandSource(this);
                }
                else
                {
                    CanExecute = true;
                    MenuItem.SetBoolField(this, BoolField.CanExecuteInvalid, true);
                }
            }
            else
            {
                CanExecute = true;
            }
        }
```

I think that we need to copy logic from WPF. It's more performant and escapes bugs on porting from WPF.
What do you think? If everyone is agree - I can provide a PR in a spare time.

### PR DESCRIPTION
**Pull request by mklts**
## What does the pull request do?
Improved CanExecute evaluation logic of MenuItem

## What is the current behavior?
CanExecute evaluation is called on any set of Command/CommandParameter even if menu is not open. this is affecting startup performance

## What is the updated/expected behavior with this PR?
CanExecute is called only on and when menu is opened. 

## Checklist
- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?

## Fixed issues
Fixes #6434

### CODE REVIEW

File: tests/Avalonia.Controls.UnitTests/MenuItemTests.cs [hunk 1]
```diff
@@ -56,7 +62,7 @@ public void MenuItem_Is_Disabled_When_Bound_Command_Doesnt_Exist()
         }
 
         [Fact]
-        public void MenuItem_Is_Disabled_When_Bound_Command_Is_Removed()
+        public void MenuItem_Is_Disabled_Bound_Command_Is_Removed()
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/MenuItemTests.cs:65]: Why the change in name here? The removal of "when" makes it grammatically incorrect.

File: tests/Avalonia.Controls.UnitTests/MenuItemTests.cs [hunk 2]
```diff
@@ -158,17 +165,160 @@ public void MenuItem_Unsubscribes_From_Command_CanExecuteChanged_When_Removed_Fr
...
 
             target.CommandParameter = true;
             Assert.True(target.IsEffectivelyEnabled);
 
             target.CommandParameter = false;
             Assert.False(target.IsEffectivelyEnabled);
         }
+        
+        [Fact]
+        public void MenuItem_Not_Invokes_CanExecute_When_ContextMenu_Closed()
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/MenuItemTests.cs:182]:
```suggestion
        public void MenuItem_Does_Not_Invoke_CanExecute_When_ContextMenu_Closed()
```

File: tests/Avalonia.Controls.UnitTests/MenuItemTests.cs [hunk 3]
```diff
@@ -158,17 +165,160 @@ public void MenuItem_Unsubscribes_From_Command_CanExecuteChanged_When_Removed_Fr
...
+                command.RaiseCanExecuteChanged();
+                Assert.Equal(3, canExecuteCallCount);
+
+                target.CommandParameter = true;
+                Assert.Equal(4, canExecuteCallCount);
+            }
+        }
+        
+        [Fact]
+        public void MenuItem_Not_Invokes_CanExecute_When_MenuFlyout_Closed()
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/MenuItemTests.cs:219]:
```suggestion
        public void MenuItem_Does_Not_Invoke_CanExecute_When_MenuFlyout_Closed()
```

File: tests/Avalonia.Controls.UnitTests/MenuItemTests.cs [hunk 4]
```diff
@@ -158,17 +165,160 @@ public void MenuItem_Unsubscribes_From_Command_CanExecuteChanged_When_Removed_Fr
...
+                command.RaiseCanExecuteChanged();
+                Assert.Equal(3, canExecuteCallCount);
+
+                target.CommandParameter = true;
+                Assert.Equal(4, canExecuteCallCount);
+            }
+        }
+        
+        [Fact]
+        public void MenuItem_Not_Invokes_CanExecute_When_Parent_MenuItem_Closed()
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/MenuItemTests.cs:257]:
```suggestion
        public void MenuItem_Does_Not_Invoke_CanExecute_When_Parent_MenuItem_Closed()
```

**Review by grokys — CHANGES_REQUESTED** [review]
Just tried this and it seems to make all menus in ControlCatalog.NetCore disabled. I think there's something wrong with the logic somehere?

**Comment by Mikolaytis** [comment]
> Just tried this and it seems to make all menus in ControlCatalog.NetCore disabled. I think there's something wrong with the logic somehere?

Looks like it is! I will check everything, add more tests, and will fix it.

**Comment by Mikolaytis** [comment]
Now it's looks like fixed. 
Also I will try to find this PR build in avalonia-all nuget, add it to our app and call our QA to check if there is any issues.
If they will find something - I will write here and provide a fix. If no messages from me for a week - all fine.

**Comment by Mikolaytis** [comment]
Our QA tested this on out app (we have lot's of menus). Did not found any issues.

**Comment by kekekeks** [comment]
@grokys

**Review by maxkatz6 — APPROVED** [review]

**Pull request merged**